### DISCUSSION TITLE
LAB 02.04 - Card trick- Task 1

### DISCUSSION THREAD

**Discussion by moralesluis26** [rramosp/ai4eng.v1#132] [02 - PYTHON]
Buenas tardes profesor y monitor. He tratado de resolver este punto pero aún no lo consigo. En la imágen se vé lo que he realizado de código, he intenté aplicar varias funciones entre ellas la de "slice" pero aún no lo consigo. Me gustaría saber si así como lo empecé está bien, y además que función adicional podría utilizar para poder llegar a realizar lo pedido. Muchas gracias de antemano. 
Código:
def make_heaps(c, n_heaps=3):
    assert n_heaps%2==1, "must have an odd number of heaps"
    assert len(c)%n_heaps==0, "the length of the deck must be a multiple of the number of heaps"
    div1 = np.array_split(c,n_heaps)
    div2=np.array(div1).tolist()
    return div2
![Captura](https://user-images.githubusercontent.com/46792756/171742653-b72ecfbe-74a5-4e10-bdc1-edd98719a826.JPG)

**Comment by Felipe-RA** [comment]
Saludos   ##moralesluis


``````
 n_heaps = 3
 c = create_deck(n_heaps=n_heaps, cards_per_heap=7)
 h = make_heaps(c, n_heaps)
 print("deck", c)
 print("heaps")
 h   

deck ['AA' 'AB' 'AC' 'AD' 'AE' 'AF' 'AG' 'AH' 'AI' 'AJ' 'AK' 'AL' 'AM' 'AN'
 'AO' 'AP' 'AQ' 'AR' 'AS' 'AT' 'AU']
heaps
[['AA', 'AD', 'AG', 'AJ', 'AM', 'AP', 'AS'],
 ['AB', 'AE', 'AH', 'AK', 'AN', 'AQ', 'AT'],
 ['AC', 'AF', 'AI', 'AL', 'AO', 'AR', 'AU']]
``````

Este es el formato apropiado que debería de seguir tu respuesta, fijate que en este caso lo que se hace es tomar una carta por heap cada `n_heap` cartas. Pero tú estás tomando las cartas de forma secuencial.

Aunque puedes crear una solución que funcione con `array_split`, la forma más directa de solucionar este problema es mediante el uso de **slicing** (ya sea en formato de función, o en su notación literal).

Recuerda la notación del slicing:

`c[start : stop : step]`

**start** :¿Desde <ins>quieres que comience tu heap?</ins>
**stop**: ¿Dónde <ins> quieres que acabe tu heap?</ins>
**step**:¿Cada  **cúantos pasos** <ins> quieres que se añada una carta a tu heap?</ins>

Una vez que entiendas cómo formar esta solución para tu primer heap puedes generalizarla al deck de cartas completo con un ciclo, o mejor aún, con un **list comprehension**

**Reply by moralesluis26** [reply]
Muchas gracias por su respuesta, ya lo pude solucionar.
Aunque el código me quedó un poco exteneso (lo hice considerando como máximo entero impar al 9).
¿La forma más corta sería usando el array_split?

**Reply by Felipe-RA** [reply]
Me alegra que lo hayas solucionado.

No te puedo decir cuál es la más corta, es más correcto que te diga la más corta que se me ocurre a mí, pues formas de solucionar un problema hay infinitas 

Yo habría echo un list slicing dentro de un  list comprehension.

Ignorando la definición de la función y los dos asserts, sale en 1 línea:

`return [list(c[i::n_heaps]) for i in range(n_heaps)]`

Lo "bonito" de esta respuesta es lo directa que se vé la relación entre los elementos que componen la línea de código y el concepto físico de crear los montones de cartas.

1. Tienes todas tus cartas en la baraja `c`. Empiezas desde la carta **i = 0** (tu <ins>start</ins> es i)
2. Quieres recorrer toda la baraja; por lo tanto no defines un**stop**
3. Vas a tomar cartas de la baraja haciendo "brincos" cada **n_heaps** cartas. Ese es tu <ins>step</ins>.
4. Cuando llegas al final de la baraja ya construiste tu primer heap!
5. El list comprehension crea un total de **n_heaps**, de ahí  el `range(n_heaps)`
6. Cuando se comienza a construir el segundo heap, el <ins>start</ins> ya es 1, entonces no vuelves a seleccionar una carta repetida en la baraja)

Esta es una forma en la que se puede hacer, por supuesto echale ganas e intenta buscar más! 

Si no te llega a quedar algo claro te puedes pasar por una sesión sincrónica y lo miramos.

PD: También estaría bien que pongas tu solución, así te puedo dar consejos!

**Reply by moralesluis26** [reply]
Saludos ##Felipe-RA

Muchas gracias por su respuesta, me has mostrado una herramienta muy interesante. Tal como lo describiste, lo estaba pensando y razonando, sòlo que no tenía claro que funciones podía usar. Por lo tanto lo hice de forma manual, quedando un código muy extenso. En principio me pareció absurdo, porque hacerlo así sólo sirve para la cantidad de n_heaps que programe, a diferencia de la forma que propones, la cual contempla cualquier nùmero siguiendo las condiciones impuestas al inicio del código. Nuevamente gracias por sus respuestas.
 
![imagen](https://user-images.githubusercontent.com/46792756/171978625-251da3be-535c-4296-92e8-e85d57ed42b0.png)
